### ISSUE TITLE
unable to install oauthd on ubuntu 14.04 with nodejs 4.1.1

### ISSUE THREAD

**Issue by rchan-pp** [oauth-io/oauthd#171]
I've installed nodejs 4.1.1 following these instructions
 https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions

oauthd fails to build with this message, with hiredis being the culprit. Not sure if this would be resolved by bumping up the hiredis version dependency, as I can see they have much newer versions available.

```
../node_modules/nan/nan.h:1975:16: error: ‘class v8::String’ has no member named ‘GetExternalAsciiStringResource’
     ext = str->GetExternalAsciiStringResource();
                ^
make: *** [Release/obj.target/hiredis/src/hiredis.o] Error 1
make: Leaving directory `/usr/lib/node_modules/oauthd/node_modules/hiredis/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 3.13.0-49-generic
gyp ERR! command "/usr/bin/nodejs" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /usr/lib/node_modules/oauthd/node_modules/hiredis
gyp ERR! node -v v4.1.1
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok

> dtrace-provider@0.2.8 install /usr/lib/node_modules/oauthd/node_modules/restify/node_modules/dtrace-provider
> node-gyp rebuild

make: Entering directory `/usr/lib/node_modules/oauthd/node_modules/restify/node_modules/dtrace-provider/build'
  TOUCH Release/obj.target/DTraceProviderStub.stamp
make: Leaving directory `/usr/lib/node_modules/oauthd/node_modules/restify/node_modules/dtrace-provider/build'
npm ERR! Linux 3.13.0-49-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install" "-g" "oauthd@1.0.0-beta.17"
npm ERR! node v4.1.1
npm ERR! npm  v2.14.4
npm ERR! code ELIFECYCLE

npm ERR! hiredis@0.1.17 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the hiredis@0.1.17 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the hiredis package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls hiredis
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ubuntu/npm-debug.log
```

**Comment by designermonkey** [comment]
+1 same error here

**Comment by DanielHuisman** [comment]
The latest version of oauthd still uses an ancient version of hiredis.
The `develop` branch has this fixed, so either wait for an updated version or manually install that branch.

**Comment by mwrzosowski** [comment]
Hi, same problem. Could you give me some tip how to install develop branch manually? So far I have been using npm installing way. Thanks a lot

**Comment by mwrzosowski** [comment]
Hi, i did install develop branch manually however can not run because of following error:

module.js:340
    throw err;
    ^

Error: Cannot find module './bin/cli/index'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:289:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/usr/local/lib/node_modules/oauthd/oauthd:4:1)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:313:12)

Any idea what could cause that???

**Comment by jjmojojjmojo** [comment]
@mwrzosowski The error happens because the oauthd script is referencing files that don't exist yet. You need to run grunt before installing. I got past that error by doing the following:

```bash
$ git clone https://github.com/oauth-io/oauthd.git
$ cd oauthd
$ git checkout develop
$ npm install
$ grunt
$ sudo npm install -g
```
Grunt will add in a bin directory in the checkout, so you can poke at bin/cli/index.js if you wanted to debug. That's the missing file.

**Comment by masaoliou** [comment]
I followed the steps from @mwrzosowski but encounter similar errors as those from OP at `npm install`:

    npm WARN package.json async@1.4.2 No README data
    npm WARN package.json ua-parser-js@0.6.2 Invalid filename in 'files' list: 
    
    > hiredis@0.4.1 install /tmp/oauthd/oauthd/node_modules/hiredis
    > node-gyp rebuild
    
    make: Entering directory '/tmp/oauthd/oauthd/node_modules/hiredis/build'
      CC(target) Release/obj.target/hiredis-c/deps/hiredis/sds.o
      CC(target) Release/obj.target/hiredis-c/deps/hiredis/read.o
      AR(target) Release/obj.target/deps/hiredis-c.a
      COPY Release/hiredis-c.a
      CXX(target) Release/obj.target/hiredis/src/hiredis.o
    In file included from ../src/reader.h:1:0,
                     from ../src/hiredis.cc:1:
    ../node_modules/nan/nan.h:324:47: error: ‘REPLACE_INVALID_UTF8’ is not a member of ‘v8::String’
       static const unsigned kReplaceInvalidUtf8 = v8::String::REPLACE_INVALID_UTF8;
                                                   ^
    hiredis.target.mk:95: recipe for target 'Release/obj.target/hiredis/src/hiredis.o' failed
    make: *** [Release/obj.target/hiredis/src/hiredis.o] Error 1
    make: Leaving directory '/tmp/oauthd/oauthd/node_modules/hiredis/build'
    gyp ERR! build error 
    gyp ERR! stack Error: `make` failed with exit code: 2
    gyp ERR! stack     at ChildProcess.onExit (/usr/share/node-gyp/lib/build.js:267:23)
    gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
    gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:809:12)
    gyp ERR! System Linux 4.6.0
    gyp ERR! command "nodejs" "/usr/bin/node-gyp" "rebuild"
    gyp ERR! cwd /tmp/oauthd/oauthd/node_modules/hiredis
    gyp ERR! node -v v0.10.29
    gyp ERR! node-gyp -v v0.12.2
    gyp ERR! not ok 
    npm WARN This failure might be due to the use of legacy binary "node"
    npm WARN For further explanations, please read
    /usr/share/doc/nodejs/README.Debian

**Comment by thyb** [comment]
We've just released the version 1.0.0-rc.1 which fix this issue by upgrading all dependencies and allow recent node version to work normally.
